### PR TITLE
Support inline tags

### PR DESCRIPTION
**Pull request by Xanderwot**
Based on [issue](https://github.com/phlex-ruby/phlex/issues/750)

Adding support of inline tag declaration
Examples:
```
h1 'Hello'
```
will render:
```
<h1>Hello</h1>
```

```
div do
  h1 'Hello', class: 'title'
end
```
will render:
```
<div><h1 class="title">Hello</h1></div>
```

### CODE REVIEW

File: lib/phlex/elements.rb [hunk 1]
```diff
@@ -59,6 +59,11 @@ def #{method_name}(**attributes, &block)
 						buffer << "<#{tag}" << (Phlex::ATTRIBUTE_CACHE[attributes] ||= __attributes__(attributes)) << ">"
 						yield_content(&block)
 						buffer << "</#{tag}>"
+					elsif content
+						buffer << "<#{tag}" << (Phlex::ATTRIBUTE_CACHE[attributes] ||= __attributes__(attributes)) << ">"
+						inline_block = content.is_a?(Proc) ? content : proc { content }
+						yield_content(&inline_block)
```
Comment (joeldrapper) [inline lib/phlex/elements.rb:65]: I think we should probably do something more like
```ruby
case content
when Block
  yield_content(&content)
when String
  plain(content)
else
  raise ArgumentError
end
```

Comment (Xanderwot) [reply lib/phlex/elements.rb]: Good catch. 
Updated. 
There is no `Block` class in ruby. all `proc` and `lambda` is a `Proc`

I moved processing method to `SGML` class. I hope it's ok. 
Tests also improved

Comment (Xanderwot) [reply lib/phlex/elements.rb]: Pls let me know if I need to update documentation somewhere. Thanks

**Review by joeldrapper — APPROVED** [review]
I like this change and it’s non-breaking. We’ve never come up with anything else to use the positional argument for so I think this makes sense. I’ve pinged @willcosgrove for his thoughts though as this is a huge change and one that would be difficult to reverse.

**Comment by willcosgrove** [comment]
@joeldrapper convinced me that this is a good change 😄

I think we should not allow nesting using this though.

```rb
h1(em("Hello")) # 💥
```

And I think wrapping it in a proc (while it does fix the issue of the elements not getting added to the buffer at the right time) is not something we should support.

**Comment by Xanderwot** [comment]
> @joeldrapper convinced me that this is a good change 😄
> 
> I think we should not allow nesting using this though.
> 
> ```ruby
> h1(em("Hello")) # 💥
> ```
> 
> And I think wrapping it in a proc (while it does fix the issue of the elements not getting added to the buffer at the right time) is not something we should support.

It's actually pretty specific cases what I faced when tests inline tags in my application. I doubt what it frequently used(`h1(em("Hello"))` I mean). Anyway proc wrapping is not a main idea. Idea of inline tags to make page structure little bit easier in some cases and not to use it in some crazy cases like huge inline tags sequence.
I did include this specific to a tests just to show how it can be solved because of ruby specific but using it delegates to a developers(if they know what they doing)
It's yours decision anyway I'm just suggesting small improvement to you(I hope so) maybe it will be useful for someone. 
Actually it comes from our migration from `fortitude` which supports this inline tags

**Comment by joeldrapper** [comment]
Okay final feedback on this:

Let's make sure we raise if you use a positional argument **_and_** a block.

```ruby
h1("Hello") { "World" }
```

Let’s make sure we raise if you pass `nil`.

```ruby
h1(h2("Hi"))
```

`h2` here returns `nil` so the `h1` call should raise. The way we can detect this is by setting the default to a different null that we can specifically check for.

Note: in order to do this, we won’t be able to use `nil` as the default (`content = nil`). Instead, let’s define a `Phlex::Null` object and use that as the default.

```ruby
module Phlex
  Null = Object.new.freeze
end
```

Then we can check against that

```
  if Phlex::Null != content
    ...
```

Let’s also remove the `Proc` option as @willcosgrove mentioned at least for now. Adding support for this later is non-breaking but we can’t remove it once we add it.

Finally, while extracting methods is generally good practice, I expect we will get better performance if we just inline the stuff in `process_inline_content` even though there will be some duplication between standard and void elements. This will save a method call for every time we render an element with positional content.

**Comment by Xanderwot** [comment]
@joeldrapper Pls take a look ;)

**Pull request merged**